Hash keys used in cached token authenticator

enj opened the pull request:
It is possible to configure the token cache to cache failures.  We
allow 1 MB of headers per request, meaning a malicious actor could
cause the cache to use a large amount of memory by filling it with
large invalid tokens.  This change hashes the token before using it
as a key.  Measures have been taken to prevent precomputation
attacks.  SHA 256 is used as the hash to prevent collisions.

Signed-off-by: Monis Khan <mkhan@redhat.com>

/kind bug

```release-note
NONE
```

@kubernetes/sig-auth-bugs @lavalamp @liggitt 

xref: #83643

enj committed:
Hash keys used in cached token authenticator

It is possible to configure the token cache to cache failures.  We
allow 1 MB of headers per request, meaning a malicious actor could
cause the cache to use a large amount of memory by filling it with
large invalid tokens.  This change hashes the token before using it
as a key.  Measures have been taken to prevent precomputation
attacks.  SHA 256 is used as the hash to prevent collisions.

Signed-off-by: Monis Khan <mkhan@redhat.com>

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -98,5 +110,17 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
+	_, _ = hash.Write(audsHash[:])
+	_, _ = hash.Write(tokenHash[:])
+
+	return base64.StdEncoding.EncodeToString(hash.Sum(nil))
```
@lavalamp did suggest that I could directly cast this to `string` since we should not be printing it.  I can make that change if others agree on it.

mikedanese replied:
+1 to casting to a string.

enj commented:
/priority important-soon

lavalamp commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -18,13 +18,25 @@ package cache
 	"k8s.io/apiserver/pkg/authentication/authenticator"
 )
 
+func init() {
```
consider storing the value as a membor of cachedTokenAuthenticator and doing this during its construction?

lavalamp commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -98,5 +110,17 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
 }
 
 func keyFunc(auds []string, token string) string {
```
A benchmark for this might be useful. I'm more worried about allocations than CPU.

lavalamp replied:
I'm also worried that `"%#v"` might print out some pointers instead of the contents; can you write a test proving that the hash value for:
* different aud pointers but same contents returns the same value?
* and: same aud pointers but different values (e.g., due to pointer reuse) returns a different value?

lavalamp commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -98,5 +110,17 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
 func keyFunc(auds []string, token string) string {
-	return fmt.Sprintf("%#v|%v", auds, token)
+	// individually hash each input to prevent ambiguities, i.e. "xy" + "z" == "x" + "yz"
+	audsHash := sha256.Sum256([]byte(fmt.Sprintf("%#v", auds)))
```
consider defining:
```
func singleDirectHash(item interface{}) []byte {
  h := sha256.New()
  fmt.Fprintf(h, "%#v", item)
  return h.Sum()
}

lavalamp replied:
That should avoid one or more allocations.

mikedanese replied:
You can avoid the re-initializing the hash and Sprintf parsing with length prefixing.

```golang
func appendString(h hash.Hash, s string) {
  b := make([]byte, 4 + len(s))
  binary.LittleEndian.PutUint32(b[0:], uint32(len(s)))
  copy(b[4:], s)
  h.Write(b) 
}
```

This solves the ambiguous plaintext problem and also requires minimal allocated bytes. You could also allocate space for all strings in slice since it's easy to calculate total size.

There's also a nice wrapper for length prefix encoding that we use in the kms plugin: https://github.com/kubernetes/kubernetes/blob/a84787465547697106f43e30e4406bb490d02670/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/envelope.go#L135-L139.

enj replied:
I went with the length prefix based approach since it resulted in the least number of allocations.

mikedanese commented:
Some discussion going on in https://github.com/kubernetes/kubernetes/issues/83375 around this. Does openshift enable the failure cache? I'm not thrilled about doing this for the successful authentication cache.

lavalamp commented:
> I'm not thrilled about doing this for the successful authentication cache.

Why not? It's unnecessary to have a bunch of valid tokens sitting around in memory. In addition to the size, even in a type-safe language I'm still uncomfortable having memory that we wouldn't want exfiltrated.

mikedanese commented:
The issue this addresses is in regard to invalid tokens. It's a lot better to say "you probably are going to get this error so here you go" than "you probably are system:master so here you go". Valid tokens (guesstimate) average 1kb on the wire so it seems unlikely that we are going to get a big boost from hashing them. I'm not convinced the extra exposure here is worth the tradeoff.

lavalamp commented:
To be clear, Mo started on this due to the size, but if you look at the code, it's really just a drastic improvement on the cache keys we use that is useful no matter whether the cache is split between the two cases or not.

mikedanese commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -98,5 +110,17 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
+	tokenHash := sha256.Sum256([]byte(fmt.Sprintf("%#v", token)))
+
+	// then hash all inputs with a random prefix
+	hash := sha256.New()
```
```
hash := "crypto/hmac".New(sha256.New, randomCachePrefix)
```
Is almost what we are doing here and is a safer API than sha256. This looks fine now but a future refactor might introduce problems.

enj replied:
Using HMAC doubled the allocations from `4` to `8` so I did not use this approach.

mikedanese replied:
This is vulnerable to length extension after the refactor, which partially undoes the dos prevention. Since this is subtle, i'm personally ok with making the gc work harder to prevent problems from creeping in. What does hmac look with a sync.Pool as @lavalamp suggested?

mikedanese replied:
This is vulnerable to length extension after the refactor, which partially undoes the dos prevention. Since this is subtle, i'm personally ok with making the gc work harder to prevent problems from creeping in. What does hmac look with a sync.Pool as @lavalamp suggested?

lavalamp replied:
Hash in a end marker or hash in the number of auds?

mikedanese commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -98,5 +110,17 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
+	hash := sha256.New()
+
+	// Write() on hash never fails
+	_, _ = hash.Write(randomCachePrefix)
```
No need for the assignment if you are going to ignore the return. However, I might check the errors just because.

mikedanese reviewed (commented):
Ya, the current construction of the cache key here is sketchy. The lower memory requirement and keeping tokens out of apiserver memory are both bonuses.

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -18,13 +18,25 @@ package cache
 )
 
+func init() {
+	if _, err := rand.Read(randomCachePrefix); err != nil {
```
`rand.Read` guarantees a complete read.

enj commented:
> Does openshift enable the failure cache?

It does not.

---

Benchmark results:

master (nil audience slice)

```
$ go test -bench '.*Key' -benchmem -v -benchtime=5s
...
BenchmarkKeyFunc-4   	20000000	       479 ns/op	     912 B/op	       2 allocs/op
PASS
ok  	k8s.io/apiserver/pkg/authentication/token/cache	10.063s
```

master (audience slice with three items):

```
$ go test -bench '.*Key' -benchmem -v -benchtime=5s
...
BenchmarkKeyFunc-4   	 3000000	      2652 ns/op	    1120 B/op	       6 allocs/op
PASS
ok  	k8s.io/apiserver/pkg/authentication/token/cache	10.629s
```

This PR (audience slice with three items, with a nil slice the only change is time with 3384 ns/op):

```
$ go test -bench '.*Key' -benchmem -v -benchtime=5s
...
BenchmarkKeyFunc-4   	 2000000	      3761 ns/op	    1088 B/op	       4 allocs/op
PASS
ok  	k8s.io/apiserver/pkg/authentication/token/cache	11.357s
```

lavalamp commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -40,6 +43,10 @@ type cachedTokenAuthenticator struct {
 	cache cache
+
+	// randomCachePrefix is a per authenticator random value that is used to prevent precomputation attacks
+	// It also mitigates hash map DOS attacks (the inputs are supplied by untrusted users)
```
DOS attacks or collision attacks?

enj replied:
Same thing?  i.e. creating keys that purposefully collide limits the usefulness of the hash map == DOS of hash map?

mikedanese replied:
Collision attacks on hash maps is a dos since it degrades hash map performance to worst case.

lavalamp replied:
Collision also means you get the attributes of whatever was in the cache, which seems worse than just a dos to me?

mikedanese replied:
It's worse in outcome but not practical given that sha256 is still considered collision resistant, and tokens should already embed sufficient entropy (e.g. service account tokens have two uuids with 122-bits of random each). This primarily protects us from collisions in the hash function used by map[] which is not collision resistant by concealing the hashing result (which becomes the key to the map[] hash function) inside the apiserver.

lavalamp replied:
Just so I'm clear, you're suggesting folks could do a birthday attack on the key hashes to find ones that collide via go map's internal hash function? And then send enough of them to make a linear search of the number of keys a significant factor in the overall request response time?

That doesn't seem like a plausible attack here (not enough tokens and/or linear is not terrible), but I like that level of paranoia.

mikedanese replied:
It's definitely paranoia :) but it's cropped up in research and CVEs since early 2000s.

https://www.usenix.org/legacy/events/sec03/tech/full_papers/crosby/crosby_html/index.html
https://github.com/google/highwayhash
https://arxiv.org/pdf/1612.06257.pdf
https://en.wikipedia.org/wiki/SipHash

lavalamp commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -70,14 +82,16 @@ func newWithClock(authenticator authenticator.Token, cacheErrs bool, successTTL,
 	auds, _ := authenticator.AudiencesFrom(ctx)
 
-	key := keyFunc(auds, token)
+	key := keyFunc(a.randomCachePrefix, auds, token)
```
optional nit: if you make keyFunc a method, you don't have to pass the prefix.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -97,6 +111,40 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
+// the length of b is assumed to be greater than or equal to the length of s + 4 (b is mutated by this function)
+func hashLengthPrefixedString(h hash.Hash, b []byte, s string) {
+	binary.BigEndian.PutUint32(b, uint32(len(s)))
+	if copy(b[4:], s) != len(s) {
```
You don't need to copy s into b? Just use two invocations of h.Write.

Then you can use a `var a [4]byte` to ensure that b is on the stack.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -97,6 +111,40 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
+	}
+
+	hashLengthPrefixedString(h, b, token)
+	for _, aud := range auds {
```
Are auds guaranteed to be in the same order every time?

I think since auds is the last thing to go in to the hash, this code is OK. If something were to be appended, though, I think we'd also want to first write in the number of auds, otherwise it'd be possible to get a collision if the last aud could be the same as that other thing.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator_test.go:
```diff
@@ -125,3 +126,24 @@ func TestCachedTokenAuthenticatorWithAudiences(t *testing.T) {
+const jwtToken = `eyJhbGciOiJSUzI1NiIsImtpZCI6IiJ9.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJvcGVuc2hpZnQtc2RuIiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZWNyZXQubmFtZSI6InNkbi10b2tlbi1nNndtYyIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VydmljZS1hY2NvdW50Lm5hbWUiOiJzZG4iLCJrdWJlcm5ldGVzLmlvL3NlcnZpY2VhY2NvdW50L3NlcnZpY2UtYWNjb3VudC51aWQi ...
+
+func BenchmarkKeyFunc(b *testing.B) {
+	randomCachePrefix := make([]byte, 32)
```
btw you can use b.Run() to make multiple related benchmarks if you get tired of manually setting up the scenarios you reported numbers for :)

lavalamp commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -97,6 +111,40 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
+	if copy(b[4:], s) != len(s) {
+		panic("invalid scratch buffer")
+	}
+	if _, err := h.Write(b[:4+len(s)]); err != nil {
```
There SHOULD be a fast way to write a string. https://golang.org/pkg/io/#WriteString comes close but sha256 doesn't support WriteString, so that's an alloc.

To get rid of this alloc I guess you'll have to do https://github.com/golang/go/issues/25484#issuecomment-391415660 :/

lavalamp commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -97,6 +111,40 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
+	}
+	b := make([]byte, 4+maxLength) // include extra bytes to store the length
+
+	h := sha256.New()
```
You can use sync.Pool + h.Reset to get rid of this alloc.

https://golang.org/pkg/sync/#Pool

lavalamp replied:
(unclear if the map lookup + lock would be faster, though)

enj commented:
I addressed the comments in a new commit, though I am not sure we want the added complexity.

```
BenchmarkKeyFunc/has_audiences-4         	 2000000	      4483 ns/op	      68 B/op	       3 allocs/op
BenchmarkKeyFunc/nil_audiences-4         	 2000000	      3827 ns/op	      68 B/op	       3 allocs/op
PASS
ok  	k8s.io/apiserver/pkg/authentication/token/cache	25.138s
```

lavalamp commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -97,6 +121,48 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
+
+// toBytes performs unholy acts to avoid allocations
+func toBytes(s string) []byte {
+	header := *(*reflect.StringHeader)(unsafe.Pointer(&s))
```
Sorry, I meant to link to the exact comment, that bug had people doing it the hard way and the easy way. Should be:

```go
func ByteSlice2String(bs []byte) string {
	return *(*string)(unsafe.Pointer(&bs))
}
```

lavalamp commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -97,6 +121,48 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
+	h.Reset()
+
+	// try to force stack allocation
+	var a [4]byte
```
nit: since this doesn't allocate, it's fine to just do it within the function, making it take fewer parameters.

enj replied:
Apparently it causes allocations when called from a `for` loop:

```
BenchmarkKeyFunc/has_audiences-4         	 2000000	      4452 ns/op	      48 B/op	       5 allocs/op
BenchmarkKeyFunc/nil_audiences-4         	 2000000	      3842 ns/op	      36 B/op	       2 allocs/op
PASS
```

lavalamp replied:
hm, we must be defeating the escape analysis :(

lavalamp commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -97,6 +121,48 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
+		writeLengthPrefixedString(h, b, aud)
+	}
+
+	key := string(h.Sum(nil)) // skip base64 encoding to save an allocation
```
Use a strings.Builder to eliminate this allocation.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -97,6 +121,48 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
+func keyFunc(hashPool *sync.Pool, auds []string, token string) string {
+	h := hashPool.Get().(hash.Hash)
+
+	h.Reset()
```
perhaps better to reset before Put()ing to prevent reuse?

enj replied:
Eh both way have issues if "done incorrectly."

lavalamp commented:
Hm, I saw one of the 3 allocations, not sure where the others are.

enj commented:
> Hm, I saw one of the 3 allocations, not sure where the others are.

![image](https://user-images.githubusercontent.com/7001223/66782552-f2143000-eea3-11e9-8796-bb8fb2a9c1c8.png)

mikedanese commented on staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:
```diff
@@ -97,6 +120,48 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
+
+	key := toString(h.Sum(nil)) // skip base64 encoding to save an allocation
+
+	hashPool.Put(h)
```
Maybe just defer this?

enj replied:
I was under the mindset of "I am doing terrible things to make this 'performant,' might as well go all the way."

lavalamp replied:
defer won't speed up the return, if that was the goal? you'd have to `go` it for that--probably not worth it over all.

lavalamp commented:
OK I think I'm done nitpicking ;)

The only pending comment that might be worth addressing is either hashing in the count of auds, or hashing in an end-marker of some sort.

fedebongio commented:
/assign @lavalamp

enj commented:
> The only pending comment that might be worth addressing is either hashing in the count of auds

Done

enj commented:
I will wait to squash until others have a chance to review.

lavalamp commented:
Looks good to me, will tag when you squash :)

mikedanese commented:
Lgtm as well

On Mon, Oct 14, 2019, 14:29 Daniel Smith <notifications@github.com> wrote:

> Looks good to me, will tag when you squash :)
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/83796?email_source=notifications&email_token=AANUKILMOEMSRV3KMYLS52LQOTQEPA5CNFSM4I76RL22YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBGTTNY#issuecomment-541931959>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANUKIJHTTMMA4BGXRT57PDQOTQEPANCNFSM4I76RL2Q>
> .
>

enj commented:
Squashed and rebased.

lavalamp commented:
/lgtm
/approve
